Handle logging in compose to swarm

vdemeester opened the pull request:
Logging configuration was completely ignore when deploy a compose file to swarm. This fixes it 👼.
This also move `Driver` in `common.go` and update the comment as it's used for network *and* logging driver.

Fixes #29116 

/cc @aanand @dnephin @thaJeztah @vieux @jpetazzo 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

cpuguy83 commented:
```
10:54:00 FAIL: docker_cli_stack_test.go:44: DockerSwarmSuite.TestStackDeployComposeFile
10:54:00 
10:54:00 [dd99d7a22589c] waiting for daemon to start
10:54:00 [dd99d7a22589c] daemon started
10:54:01 docker_cli_stack_test.go:55:
10:54:01     c.Assert(err, checker.IsNil, check.Commentf(out))
10:54:01 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc42072afa0), Stderr:[]uint8(nil)} ("exit status 2")
10:54:01 ... Creating network testdeploy_default
10:54:01 panic: runtime error: invalid memory address or nil pointer dereference
10:54:01 [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x670d30]
10:54:01 
10:54:01 goroutine 1 [running]:
10:54:01 panic(0xb1bea0, 0xc420014050)
10:54:01 	/usr/local/go/src/runtime/panic.go:500 +0x1a1
10:54:01 github.com/docker/docker/cli/command/stack.convertService(0x7ffc74e48bd0, 0xa, 0xc4203ca0a8, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
10:54:01 	/go/src/github.com/docker/docker/cli/command/stack/deploy.go:593 +0x770
10:54:01 github.com/docker/docker/cli/command/stack.convertServices(0x7ffc74e48bd0, 0xa, 0xc42005dbc0, 0x7ffc74e48bd0, 0xa, 0xc4201965d0)
10:54:01 	/go/src/github.com/docker/docker/cli/command/stack/deploy.go:515 +0x16f
10:54:01 github.com/docker/docker/cli/command/stack.deployCompose(0x116fd00, 0xc420015568, 0xc4201b7560, 0x0, 0x0, 0x7ffc74e48bb3, 0x1c, 0x7ffc74e48bd0, 0xa, 0x0, ...)
10:54:01 	/go/src/github.com/docker/docker/cli/command/stack/deploy.go:134 +0x60e
10:54:01 github.com/docker/docker/cli/command/stack.runDeploy(0xc4201b7560, 0x0, 0x0, 0x7ffc74e48bb3, 0x1c, 0x7ffc74e48bd0, 0xa, 0x0, 0x0, 0xc4203464a0)
10:54:01 	/go/src/github.com/docker/docker/cli/command/stack/deploy.go:74 +0x117
10:54:01 github.com/docker/docker/cli/command/stack.newDeployCommand.func1(0xc4203b6d80, 0xc4203b2bd0, 0x1, 0x3, 0x0, 0x0)
10:54:01 	/go/src/github.com/docker/docker/cli/command/stack/deploy.go:52 +0x93
10:54:01 github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc4203b6d80, 0xc42000c0d0, 0x3, 0x3, 0xc4203b6d80, 0xc42000c0d0)
10:54:01 	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:646 +0x26d
10:54:01 github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42007f8c0, 0xc42007fd40, 0xc42032cdf0, 0xc420026450)
10:54:01 	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:742 +0x377
10:54:01 github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42007f8c0, 0xc42007f8c0, 0x1167940)
10:54:01 	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:695 +0x2b
10:54:01 main.main()
10:54:01 	/go/src/github.com/docker/docker/cmd/docker/docker.go:104 +0xcb
10:54:01 
10:54:01 goroutine 17 [syscall, locked to thread]:
10:54:01 runtime.goexit()
10:54:01 	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1
10:54:01 
10:54:01 goroutine 5 [syscall]:
10:54:01 os/signal.signal_recv(0x0)
10:54:01 	/usr/local/go/src/runtime/sigqueue.go:116 +0x157
10:54:01 os/signal.loop()
10:54:01 	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
10:54:01 created by os/signal.init.1
10:54:01 	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41
10:54:01 
10:54:01 goroutine 9 [IO wait]:
10:54:01 net.runtime_pollWait(0x7f9c040a9fb8, 0x72, 0x4)
10:54:01 	/usr/local/go/src/runtime/netpoll.go:160 +0x59
10:54:01 net.(*pollDesc).wait(0xc4203be8b0, 0x72, 0xc4200379d0, 0xc420014150)
10:54:01 	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
10:54:01 net.(*pollDesc).waitRead(0xc4203be8b0, 0x116a940, 0xc420014150)
10:54:01 	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
10:54:01 net.(*netFD).Read(0xc4203be850, 0xc4203c6000, 0x1000, 0x1000, 0x0, 0x116a940, 0xc420014150)
10:54:01 	/usr/local/go/src/net/fd_unix.go:243 +0x1a1
10:54:01 net.(*conn).Read(0xc420026598, 0xc4203c6000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
10:54:01 	/usr/local/go/src/net/net.go:173 +0x70
10:54:01 net/http.(*persistConn).Read(0xc420291000, 0xc4203c6000, 0x1000, 0x1000, 0x5ae780, 0xc420037b58, 0x4069cd)
10:54:01 	/usr/local/go/src/net/http/transport.go:1261 +0x154
10:54:01 bufio.(*Reader).fill(0xc4203c2300)
10:54:01 	/usr/local/go/src/bufio/bufio.go:97 +0x10c
10:54:01 bufio.(*Reader).Peek(0xc4203c2300, 0x1, 0xc420037bbd, 0x1, 0x0, 0xc420016e40, 0x0)
10:54:01 	/usr/local/go/src/bufio/bufio.go:129 +0x62
10:54:01 net/http.(*persistConn).readLoop(0xc420291000)
10:54:01 	/usr/local/go/src/net/http/transport.go:1418 +0x1a1
10:54:01 created by net/http.(*Transport).dialConn
10:54:01 	/usr/local/go/src/net/http/transport.go:1062 +0x4e9
10:54:01 
10:54:01 goroutine 10 [select]:
10:54:01 net/http.(*persistConn).writeLoop(0xc420291000)
10:54:01 	/usr/local/go/src/net/http/transport.go:1646 +0x3bd
10:54:01 created by net/http.(*Transport).dialConn
10:54:01 	/usr/local/go/src/net/http/transport.go:1063 +0x50e
10:54:01 
10:54:01 goroutine 13 [runnable]:
10:54:01 github.com/docker/docker/vendor/golang.org/x/net/context/ctxhttp.Do.func3(0x116fd00, 0xc420015568, 0xc4200e1890, 0xc420377f80)
10:54:01 	/go/src/github.com/docker/docker/vendor/golang.org/x/net/context/ctxhttp/ctxhttp.go:76
10:54:01 created by github.com/docker/docker/vendor/golang.org/x/net/context/ctxhttp.Do
10:54:01 	/go/src/github.com/docker/docker/vendor/golang.org/x/net/context/ctxhttp/ctxhttp.go:83 +0x2d2
10:54:01 
10:54:01 
10:54:01 [dd99d7a22589c] exiting daemon
```

vdemeester commented:
@cpuguy83 ahah, I forgot to put a check somewhere :joy:

vdemeester committed:
Handle logging in compose to swarm

Logging configuration was completely ignore when deploy a compose file
to swarm. This fixes it.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

dnephin reviewed (approved):
LGTM

Good catch

cpuguy83 reviewed (approved):
LGTM

yaakov-berkovitch commented:
i'm using 1.13 rc4 and this issue is not fix. Is there any version this issue is fix and i can install ?

thaJeztah commented:
@yaakov-berkovitch could you open an issue with more details (possibly a simple compose file to reproduce)?

yaakov-berkovitch commented:
Following the compose-file I'm using:

```
version: '3'

services:
  hello-world-daemon:
    labels: [app.container=hello-world-daemon]
    deploy:
      placement:
        constraints: [app.container != hello-world-daemon]
      labels: [app.service=hello-world-daemon]
      replicas: 1
      restart_policy:
        condition: on-failure
        delay: 20s
        max_attempts: 10
    image: hello-world-daemon:latest
    logging:
      driver: gelf
      options:
        gelf-address: udp://147.234.150.167:5555

```

The output of the docker logs commands for my container:
    "logs" command is supported only for "json-file" and "journald" logging drivers (got: gelf)

Docker release:
Client:
 Version:      1.13.0-rc4
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Fri Dec 16 22:55:47 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc4
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Fri Dec 16 22:55:47 2016
 OS/Arch:      linux/amd64
 Experimental: true

Thanks.

thaJeztah commented:
@yaakov-berkovitch but that's the expected behavior; you specified the "gelf" logging driver for your service. The Gelf logging driver sends all logs to a remote server, which means that docker cannot read back those logs (through the `docker logs` command). That's exactly what this issue fixed; previously the logging driver was ignored, resulting in the default (json-file) logging driver to be used.

You can see in the reference docs for the `docker logs` command that reading/viewing logs is not supported when using the gelf logging driver; only when using the json-file or journald logging drivers; https://docs.docker.com/engine/reference/commandline/logs/

yaakov-berkovitch commented:
Sorry for your time, you are correct. All looks fine and it's working.
Thanks for your explanation.

yaakov-berkovitch commented:
Don't know if it's an issue, but I just moved to docker 1.13 rc5, and I got this error:
`Error": "Failed to initialize logging driver: gelf: cannot connect to GELF endpoint: epic_logstash:12201 dial udp: lookup epic_logstash on 147.234.244.5:53: server misbehaving",
`My setup is : one dockerized application and dockerized logstash running on overlay network.
BTW, i have no idea what is the address 147.234.244.5. Looks like inside my company networks, but not my VM.

Thanks for any tips.
